### PR TITLE
Fix for --disable-deprecated (#1273)

### LINKED ISSUE #1273: fix --disable-deprecated
I am trying to build a minimal libvips version for Debian (only supporting web formats), so I used `./configure --disable-deprecated`. Imagemagick is not installed.

I am using 8.7.4, and ran `./configure --disable-deprecated && make`.

The compilation fails when building tools:
```
/bin/bash ../libtool  --tag=CC   --mode=link gcc  -O3   -o vips vips.o -DG_DISABLE_ASSERT -DG_DISABLE_CHECKS -pthread -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/x86_64-linux-gnu -I/usr/include/pango-1.0 -I/usr/include/orc-0.4 -I/usr/include/libpng16 -I/usr/include/libexif -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/include/freetype2 ../libvips/libvips.la -lz  -lpng16  -ltiff -ljpeg  -Wl,--export-dynamic -lgmodule-2.0 -pthread -lgobject-2.0 -lglib-2.0  -lexpat -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype   -lorc-0.4 -llcms2 -lgif        -lwebp -lwebpmux -lwebp  -lexif -lm
libtool: link: gcc -O3 -o .libs/vips vips.o -DG_DISABLE_ASSERT -DG_DISABLE_CHECKS -pthread -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/x86_64-linux-gnu -I/usr/include/pango-1.0 -I/usr/include/orc-0.4 -I/usr/include/libpng16 -I/usr/include/libexif -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/include/freetype2 -Wl,--export-dynamic -pthread  ../libvips/.libs/libvips.so -lz -lpng16 -ltiff -ljpeg -lgmodule-2.0 /usr/lib/x86_64-linux-gnu/libexpat.so -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig /usr/lib/x86_64-linux-gnu/libfreetype.so -lorc-0.4 -llcms2 -lgif -lwebpmux -lwebp -lexif -lm -pthread
vips.o: In function `print_links':
vips.c:(.text+0x11e): undefined reference to `im_map_packages'
vips.o: In function `print_list':
vips.c:(.text+0x8b5): undefined reference to `im_find_package'
vips.c:(.text+0x972): undefined reference to `im_map_packages'
vips.c:(.text+0x9b5): undefined reference to `im_map_packages'
vips.c:(.text+0x9bf): undefined reference to `im_find_function'
vips.o: In function `print_cppdecls':
vips.c:(.text+0x115c): undefined reference to `im_find_package'
vips.c:(.text+0x11fd): undefined reference to `im_map_packages'
vips.c:(.text+0x1207): undefined reference to `im_find_function'
vips.o: In function `print_cppdefs':
vips.c:(.text+0x24dc): undefined reference to `im_find_package'
vips.c:(.text+0x257d): undefined reference to `im_map_packages'
vips.c:(.text+0x2587): undefined reference to `im_find_function'
vips.o: In function `main':
vips.c:(.text.startup+0x1bf): undefined reference to `im_load_plugin'
vips.c:(.text.startup+0x30b): undefined reference to `im_find_function'
vips.c:(.text.startup+0x32f): undefined reference to `im_run_command'
vips.c:(.text.startup+0x34a): undefined reference to `im_package_of_function'
../libvips/.libs/libvips.so: undefined reference to `im_stats'
../libvips/.libs/libvips.so: undefined reference to `vips_check_coding_rad'
../libvips/.libs/libvips.so: undefined reference to `im_open'
../libvips/.libs/libvips.so: undefined reference to `im_matinv'
../libvips/.libs/libvips.so: undefined reference to `im_dvector'
../libvips/.libs/libvips.so: undefined reference to `im_dmat_alloc'
../libvips/.libs/libvips.so: undefined reference to `im_Lab2disp'
../libvips/.libs/libvips.so: undefined reference to `im_close_plugins'
../libvips/.libs/libvips.so: undefined reference to `vips__affine'
../libvips/.libs/libvips.so: undefined reference to `vips_info_set'
../libvips/.libs/libvips.so: undefined reference to `im_scale'
../libvips/.libs/libvips.so: undefined reference to `im_embed'
../libvips/.libs/libvips.so: undefined reference to `im_create_dmask'
../libvips/.libs/libvips.so: undefined reference to `im_load_plugins'
../libvips/.libs/libvips.so: undefined reference to `im_mattrn'
../libvips/.libs/libvips.so: undefined reference to `im_demand_hint'
../libvips/.libs/libvips.so: undefined reference to `vips_check_imask'
../libvips/.libs/libvips.so: undefined reference to `im_local_dmask'
../libvips/.libs/libvips.so: undefined reference to `im__format_init'
../libvips/.libs/libvips.so: undefined reference to `im_extract_area'
../libvips/.libs/libvips.so: undefined reference to `im_maxpos'
../libvips/.libs/libvips.so: undefined reference to `im_close'
../libvips/.libs/libvips.so: undefined reference to `im_notequalconst'
../libvips/.libs/libvips.so: undefined reference to `im_cp_desc'
../libvips/.libs/libvips.so: undefined reference to `im_add_callback'
../libvips/.libs/libvips.so: undefined reference to `im_LabQ2LabS'
../libvips/.libs/libvips.so: undefined reference to `im_free_dmat'
../libvips/.libs/libvips.so: undefined reference to `im_maplut'
../libvips/.libs/libvips.so: undefined reference to `im_copy'
../libvips/.libs/libvips.so: undefined reference to `im_andimage'
../libvips/.libs/libvips.so: undefined reference to `im_clip2fmt'
../libvips/.libs/libvips.so: undefined reference to `im_create_dmaskv'
../libvips/.libs/libvips.so: undefined reference to `im_LabQ2Lab'
../libvips/.libs/libvips.so: undefined reference to `im_vips2imask'
../libvips/.libs/libvips.so: undefined reference to `im_free_dmask'
../libvips/.libs/libvips.so: undefined reference to `im_cp_descv'
../libvips/.libs/libvips.so: undefined reference to `im_spcor'
../libvips/.libs/libvips.so: undefined reference to `im_identity_ushort'
../libvips/.libs/libvips.so: undefined reference to `im_avg'
../libvips/.libs/libvips.so: undefined reference to `im_identity'
../libvips/.libs/libvips.so: undefined reference to `im_free_imask'
../libvips/.libs/libvips.so: undefined reference to `im_generate'
../libvips/.libs/libvips.so: undefined reference to `im_ifthenelse'
../libvips/.libs/libvips.so: undefined reference to `im_powtra'
../libvips/.libs/libvips.so: undefined reference to `im_dup_imask'
../libvips/.libs/libvips.so: undefined reference to `im_open_local'
../libvips/.libs/libvips.so: undefined reference to `im_black'
../libvips/.libs/libvips.so: undefined reference to `im_extract_band'
../libvips/.libs/libvips.so: undefined reference to `im_matmul'
../libvips/.libs/libvips.so: undefined reference to `im_local_imask'
../libvips/.libs/libvips.so: undefined reference to `im_open_local_array'
../libvips/.libs/libvips.so: undefined reference to `im_invmat'
../libvips/.libs/libvips.so: undefined reference to `im_lintra'
../libvips/.libs/libvips.so: undefined reference to `im_insert'
../libvips/.libs/libvips.so: undefined reference to `im_free_dvector'
../libvips/.libs/libvips.so: undefined reference to `im_skip_dir'
../libvips/.libs/libvips.so: undefined reference to `im__insert_base'
collect2: error: ld returned 1 exit status
Makefile:618: recipe for target 'vips' failed
make[2]: *** [vips] Error 1
make[2]: Leaving directory '/root/vips-8.7.4/tools'
```

This is non-blocking as I only need `libvips`, so I can run `make && make install` in `vips-8.7.4/libvips`, but it would be great to be able to build the tools with `--disable-deprecated`.

Note: the issue was originally saying it was because IM was not installed, I updated it as it is indeed about `--disable-deprecated`

### PR DESCRIPTION
**Pull request by kleisauke**
~~Marked this PR as draft since the whole mosaicing folder has yet to be converted to vips8.~~

This PR resolves #1273 and #798.

### CODE REVIEW

**Comment by kleisauke** [comment]
This PR is almost done, the whole mosaicing folder was converted to vips8 with commit https://github.com/libvips/libvips/pull/1593/commits/2105b76799d7c8f10b2b46cf45e2d8b7dc4125a1. I'm still working out a few kinks, but feel free to review/comment. Possible further improvements has been preceded by `TODO(kleisauke):`.

There's now a unit test for `vips_mosaic` and `vips_globalbalance` (ported from the 1 point mosaic nip2 demo). Although  `vips_globalbalance` seems to [output correctly](https://user-images.githubusercontent.com/12746591/79687530-28ae7700-8248-11ea-87ed-4eceb2ae05b3.jpg) for `int_output=True` [the float output](https://user-images.githubusercontent.com/12746591/79687451-9c03b900-8247-11ea-811d-84aaa5f88c50.jpg) however seems to be wrong.

I also added a new vips8 class named `vips_inversematrix` (perhaps it should be named `vips_invertmatrix`(?)) for inverting square matrices. This class has been converted from `im_matinv`. The logic for transposing and multiplying matrices are small, so these functions (named `vips__transposematrix` and `vips__multiplymatrix`) are included in `global_balance.c`.

**Comment by kleisauke** [comment]
> Although `vips_globalbalance` seems to [output correctly](https://user-images.githubusercontent.com/12746591/79687530-28ae7700-8248-11ea-87ed-4eceb2ae05b3.jpg) for `int_output=True` [the float output](https://user-images.githubusercontent.com/12746591/79687451-9c03b900-8247-11ea-811d-84aaa5f88c50.jpg) however seems to be wrong.

Fixed with https://github.com/libvips/libvips/pull/1593/commits/5b1b8cd8a99b68eea074d3b710f282b22729c2c9.

**Comment by kleisauke** [comment]
@jcupitt Commit https://github.com/libvips/libvips/pull/1593/commits/9cf85883be4a6c9b37d75711bc801452a88006a0 adds a couple of TODOs outside the scope of this PR. For example, it looks like `affine->interpolate` leaks when `vips_affine` is called without an interpolator:
https://github.com/libvips/libvips/blob/1ef3594be122da22454b9e68eb8704b7d921df18/libvips/resample/affine.c#L426-L429
An easy way to fix this would be to unref the interpolator on close, i.e.:
```diff
diff --git a/libvips/resample/affine.c b/libvips/resample/affine.c
index 1111111..2222222 100644
--- a/libvips/resample/affine.c
+++ b/libvips/resample/affine.c
@@ -425,8 +425,10 @@ vips_affine_build( VipsObject *object )
 
 	/* Can be set explicitly to NULL to mean default setting. 
 	 */
-	if( !affine->interpolate )
+	if( !vips_object_argument_isset( object, "interpolate" ) ) {
 		affine->interpolate = vips_interpolate_new( "bilinear" );
+		vips_object_local( object, affine->interpolate );
+	}
 
 	in = resample->in;
 
```
(untested)

**Comment by kleisauke** [comment]
Ignore my previous comment, it looks like non-assigned inputs (i.e. input that is not explicitly assigned by the user) are automatically finalized, TIL. So for example the `vips_object_local` here:
https://github.com/libvips/libvips/blob/e1baf66f194cd0ee3d2cf66d41e1fc12e2d597be/libvips/resample/resize.c#L267-L271

Is correct, but not necessary if you do this in `vips_affine_build`.

**Comment by jcupitt** [comment]
Hello Kleis, yes, I think the verb form is usually better, so `invertmatrix`.

Actually, we have `matrixload`, so perhaps `matrixinvert` might be better?

**Comment by kleisauke** [comment]
OK, I cleaned up the commits and renamed it to `matrixinvert` with https://github.com/kleisauke/libvips/commit/45f9999e7e30a991de7c39e3970590493a476f53.

This is now ready for reviewing. The mosaic changes are very large, so I think it's wise to fuzz it for a few days before the 8.10 release.

As an aside, I noticed that the `test_descriptors` with `make check` is currently failing when building with libspng.
```
** seq open ..
** crop1, avg ..
** crop2, avg ..
** unref ..
** shutdown ..
lt-test_descriptors: /home/kleisauke/libvips/test/test-suite/images/sample.png: fd not closed after first read
** seq open ..
** crop1, avg ..
FAIL test_descriptors.sh (exit status: 1)
```

**Comment by jcupitt** [comment]
Ooop, you're fight, it was missing minimise support. It should be fixed now.

Well done on getting all this refactoring sorted out Kleis! I'll try to read it all next week, then perhaps we can aim for July for the 8.10 release.

File: libvips/iofuncs/init.c [hunk 1]
```diff
@@ -312,6 +315,32 @@ set_stacksize( guint64 size )
 #endif /*HAVE_PTHREAD_DEFAULT_NP*/
 }
 
+static void
+vips_verbose( void ) 
+{
+	/* Older glibs were showing G_LOG_LEVEL_{INFO,DEBUG} messages
+	 * by default
+	 */
+#if GLIB_CHECK_VERSION ( 2, 31, 0 )
```
Comment (jcupitt) [inline libvips/iofuncs/init.c:324]: Oh, nice!

File: libvips/mosaicing/global_balance.c [hunk 2]
```diff
@@ -1418,73 +1459,137 @@ im__build_mosaic( SymbolTable *st, IMAGE *out, transform_fn tfn, void *a )
...
-		error_exit( "internal error #982369824375987" );
+		vips_error_exit( "internal error #982369824375987" );
 		/*NOTEACHED*/
 	}
 
 	return( 0 );
 }
 
+static int
+vips__transposematrix( VipsImage *in, VipsImage **out )
```
Comment (jcupitt) [inline libvips/mosaicing/global_balance.c:1486]: Perhaps should be matrixtranspose, though it doesn't really matter in a static.

**Comment by jcupitt** [comment]
OK, I read the whole thing, nice work! You found a lot of useful small fixes too, and even fixed some spelling.

I wonder how useful some of that old mosaicing stuff is. Nicos wrote most it back in the early 90s. Perhaps it should be moved out of the core library? That's another question I suppose.

Anyway, it looks great, let's merge.

**Comment by jcupitt** [comment]
I moved matrixinvert to mosaicing (`create/` was supposed to just be things that create images), fixed a small memleak, and added a note about your achievement to the main changelog.

**Comment by kleisauke** [comment]
Thanks! The mosaicing stuff could still be useful for museums, but I'm not sure if it's often used. Perhaps [Operation Night Watch](https://www.rijksmuseum.nl/en/nightwatch) uses this?

Could the `vips_matrixinvert_get_type` call also be moved to the [`vips_mosaicing_operation_init`](https://github.com/libvips/libvips/blob/5540fa97f4952d9bba1bd675a40e4fdd574b4ead/libvips/mosaicing/mosaicing.c#L96) function?
https://github.com/libvips/libvips/blob/e3a500204a5b2ca6e491c78d8f56f9622b7f3f06/libvips/create/create.c#L172

The `POTFILES.in` needs to be regenerated too.
https://github.com/libvips/libvips/blob/5540fa97f4952d9bba1bd675a40e4fdd574b4ead/po/POTFILES.in#L212

(It should be much easier with [these instructions](https://github.com/libvips/libvips/blob/5540fa97f4952d9bba1bd675a40e4fdd574b4ead/po/README#L24-L37) now)

**Comment by jcupitt** [comment]
OK, done!

**Comment by kleisauke** [comment]
I had a go to check the ABI compatibility after this PR, see the `abi-compliance-checker` result:
https://kleisauke.nl/compat_reports/vips/8.9.2_to_master/compat_report.html

I guess we still have to write some vips7 compat wrappers for the `im_correl`, `im_lrmosaic`, `im_tbmosaic`, `im_lrmerge` and `im_tbmerge` functions in `deprecated/vips7compat.c`? I could pick this up, if you want.

**Comment by jcupitt** [comment]
If you have time, that would be great!

We'll need to make `vips_lrmerge()` etc. more vips8-y too I think, or there will be API breaks when (if) we eventually make them into classes too. They'll need at least the `..., NULL)` parameter ending.

**Comment by kleisauke** [comment]
OK, I started to rework `vips_correl` as vips8 class, see: https://github.com/libvips/libvips/compare/master...kleisauke:mosaicing-vips8.

I'm not sure if I should rework `im_{lr,tb}mosaic` and `im_{lr,tb}merge` as vips8 classes, since `vips_mosaic` and `vips_merge` are already there. Should I try to merge these left-right/top-bottom functions within `vips_merge`/`vips_mosaic`? The compat layer can then call up these functions with the appropriate `VipsDirection`.

**Comment by jcupitt** [comment]
Ah I'd actually forgotten about them, heh. That was a while ago.

Yes, you're right, `vips_merge()` is the vips8 class interface to the old `im_lrmerge()` and `im_tbmerge()`. We should rename your `vips_lrmerge()` as `vips__lrmerge()` and have it as internal API. As you say, we could even fold them into `merge.c`, though that would make the file very large.

Shall I do this?

I wonder if `correl` is very useful. Perhaps it should just be internal API. If we make it an official part of vips8, we'll have to support the stupid thing for all time.

**Comment by jcupitt** [comment]
Off-topic: I had a go at a first draft of "what's new in 8.10":

https://github.com/libvips/libvips/blob/whatsnew-8.10/_posts/2020-06-18-What's-new-in-8.10.md

Any comments very welcome etc.

**Comment by kleisauke** [comment]
I agree, lets make `vips_correl`, `vips_{lr,tb}mosaic` and `vips_{lr,tb}merge` part of the internal API. `vips_correl` as a vips8 class will also lead to confusion since there is already a `vips_correlation` function.

I've already prepared the vips7 compat wrapper functions, see: https://github.com/libvips/libvips/compare/master...kleisauke:mosaicing-vips8.

ABI compatibility looks OK now, only some internal API symbols have been removed.
https://kleisauke.nl/compat_reports/vips/8.9.2_to_mosaicing-vips8/compat_report.html

Feel free to mark those mosaicing functions as internal, I can do a rebase after that.

**Comment by kleisauke** [comment]
The "what's new in 8.10" news post looks great! The only thing I noticed is that libspng is mentioned twice in the ChangeLog, see:
https://github.com/libvips/libvips/blob/2b2fabcd3be1d850653e09571e68d2cdb9621a80/ChangeLog#L22
https://github.com/libvips/libvips/blob/2b2fabcd3be1d850653e09571e68d2cdb9621a80/ChangeLog#L31

**Comment by jcupitt** [comment]
OK, I'll shift them to internal-only. Thanks for the note about the changelog, I've fixed that too.

**Comment by randy408** [comment]
If the Windows release will default to libspng for loading PNGs it would mean a lot if it was mentioned in the changelog or blogpost.

**Comment by jcupitt** [comment]
Good idea @randy408, I'll add a note.

**Pull request merged**